rasterio/sample.py

```diff
@@ -57,13 +94,23 @@ def sample_gen(dataset, xy, indexes=None, masked=False):
         mask = [MaskFlags.all_valid not in dataset.mask_flag_enums[i-1] for i in indexes]
         nodata = np.ma.array(nodata, mask=mask)
 
-    for pts in _grouper(xy, 256):
-        pts = zip(*filter(None, pts))
+    # Intermediate conversion to row/col coordinates
+    samples = transform_xy(dataset, xy)
```

Comment:
@groutr this isn't what I'm looking for. I would like *all* the sorting issues to be outside the function. And no new keyword argument. I think that if we want to benefit from sorting, then `xy` should be sorted pairs.